### PR TITLE
Fix SEGV in simulation with default `cable_cell`

### LINKED ISSUE #1540: Empty Cable leads to SegFault in Simulation::run
**Describe the bug**
Running this minimal example
```C++
﻿#include <arbor/recipe.hpp>
#include <arbor/simulation.hpp>
#include <arbor/cable_cell.hpp>
#include <arbor/context.hpp>
#include <arbor/load_balance.hpp>

struct recipe: arb::recipe {
    arb::cell_size_type n = 1;
    arb::cell_kind kind = arb::cell_kind::cable;
    std::vector<arb::probe_info> probes;
    arb::cable_cell_global_properties properties;
    arb::cable_cell cell;

    arb::cell_size_type num_cells() const override { return n; }
    arb::cell_kind get_cell_kind(arb::cell_gid_type) const override { return kind; }
    arb::util::unique_any get_cell_description(arb::cell_gid_type) const override { return {cell}; }
    std::vector<arb::probe_info> get_probes(arb::cell_gid_type gid) const override { return probes; }
    std::any get_global_properties(arb::cell_kind) const override { return properties; }
};

int main() {
    auto r = recipe{};
    r.properties.default_parameters = arb::neuron_parameter_defaults;
    auto c = arb::make_context();
    auto d = arb::partition_load_balance(r, c);
    auto s = arb::simulation(r, d, c);
    s.run(100, 0.01);
}
```
leads to a segmentation violation in the matrix solver.
Example debugger output
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
    frame #0: 0x00000001000807a2 test-arb-null`arb::multicore::matrix_state<double, int>::matrix_state(std::__1::vector<int, std::__1::allocator<int> > const&, std::__1::vector<int, std::__1::allocator<int> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<int, std::__1::allocator<int> > const&) + 5106
test-arb-null`arb::multicore::matrix_state<double, int>::matrix_state:
->  0x1000807a2 <+5106>: vmovsd (%rsi,%rdi,8), %xmm1      ; xmm1 = mem[0],zero
    0x1000807a7 <+5111>: vxorpd %xmm0, %xmm1, %xmm2
    0x1000807ab <+5115>: vmovlpd %xmm2, (%rdx,%rdi,8)
    0x1000807b0 <+5120>: vaddsd (%rax,%rdi,8), %xmm1, %xmm2
```
and a backtrace
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
  * frame #0: 0x00000001000807a2 test-arb-null`arb::multicore::matrix_state<double, int>::matrix_state(std::__1::vector<int, std::__1::allocator<int> > const&, std::__1::vector<int, std::__1::allocator<int> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<int, std::__1::allocator<int> > const&) + 5106
    frame #1: 0x000000010007f1fa test-arb-null`arb::matrix<arb::multicore::backend, arb::multicore::matrix_state<double, int> >::matrix(std::__1::vector<int, std::__1::allocator<int> > const&, std::__1::vector<int, std::__1::allocator<int> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<int, std::__1::allocator<int> > const&) + 1962
    frame #2: 0x0000000100078414 test-arb-null`arb::fvm_lowered_cell_impl<arb::multicore::backend>::initialize(std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, arb::recipe const&, std::__1::vector<int, std::__1::allocator<int> >&, std::__1::vector<arb::target_handle, std::__1::allocator<arb::target_handle> >&, arb::probe_association_map&) + 2852
    frame #3: 0x0000000100092cb5 test-arb-null`arb::mc_cell_group::mc_cell_group(std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, arb::recipe const&, std::__1::unique_ptr<arb::fvm_lowered_cell, std::__1::default_delete<arb::fvm_lowered_cell> >) + 1029
    frame #4: 0x000000010004e1d9 test-arb-null`std::__1::__function::__func<arb::cell_kind_implementation(arb::cell_kind, arb::backend_kind, arb::execution_context const&)::$_2, std::__1::allocator<arb::cell_kind_implementation(arb::cell_kind, arb::backend_kind, arb::execution_context const&)::$_2>, std::__1::unique_ptr<arb::cell_group, std::__1::default_delete<arb::cell_group> > (std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, arb::recipe const&)>::operator()(std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, arb::recipe const&) + 89
    frame #5: 0x00000001000ebcb4 test-arb-null`std::__1::__function::__func<arb::threading::task_group::wrap<void arb::threading::parallel_for::apply<void arb::simulation_state::foreach_group_index<arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context)::$_1>(arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context)::$_1&&)::'lambda'(int)>(int, int, arb::threading::task_system*, arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context)::$_1)::'lambda'()>, std::__1::allocator<arb::threading::task_group::wrap<void arb::threading::parallel_for::apply<void arb::simulation_state::foreach_group_index<arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context)::$_1>(arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context)::$_1&&)::'lambda'(int)>(int, int, arb::threading::task_system*, arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context)::$_1)::'lambda'()> >, void ()>::operator()() + 132
    frame #6: 0x00000001000fa407 test-arb-null`arb::threading::task_system::try_run_task() + 311
    frame #7: 0x00000001000e7919 test-arb-null`arb::threading::task_group::wait() + 41
    frame #8: 0x00000001000e625d test-arb-null`arb::simulation_state::simulation_state(arb::recipe const&, arb::domain_decomposition const&, arb::execution_context) + 1805
    frame #9: 0x00000001000e8803 test-arb-null`arb::simulation::simulation(arb::recipe const&, arb::domain_decomposition const&, std::__1::unique_ptr<arb::execution_context, arb::execution_context_deleter> const&) + 147
    frame #10: 0x00000001000042e4 test-arb-null`main at test-arb-null.cxx:26:14
```

### PR DESCRIPTION
**Pull request by thorstenhater**
Closes #1540.

- Relevant corner cases are skipped in the matrix solver.
- regression test added 
- `make_span` cleaned up a bit
- tests for `span(a + n, a)` added.

### CODE REVIEW

File: arbor/backends/multicore/matrix_state.hpp [hunk 1]
```diff
@@ -116,7 +118,7 @@ struct matrix_state {
         for (auto cv_span: util::partition_view(cell_cv_divs)) {
             auto first = cv_span.first;
             auto last = cv_span.second; // one past the end
-
+            if (first >= last) continue; // skip empty / reversed partitions. Bugs lurk.
```
Comment (bcumming) [inline arbor/backends/multicore/matrix_state.hpp:121]: Instead of the comment referring to the
```C++
if (first >= last) continue; // skip cell with no CVs
```

File: arbor/util/span.hpp [hunk 2]
```diff
@@ -22,24 +22,24 @@ namespace util {
...
-template <typename I, typename J>
+template <typename I, typename J, typename = std::enable_if_t<std::is_integral<std::common_type_t<I, J>>::value>>
 span<std::common_type_t<I, J>> make_span(I left, J right) {
     return span<std::common_type_t<I, J>>(left, right);
 }
 
 template <typename I, typename J>
 span<std::common_type_t<I, J>> make_span(std::pair<I, J> interval) {
-    return span<std::common_type_t<I, J>>(interval.first, interval.second);
+    return make_span(interval.first, interval.second);
```
Comment (bcumming) [inline arbor/util/span.hpp:32]: Are we certain that the type of `make_span` will match `span<std::common_type_t<I, J>>`?
And why are we not now testing for `std::is_integral`?

Same question for the uses below.

File: test/unit/test_simulation.cpp [hunk 3]
```diff
@@ -35,6 +38,24 @@ static auto n_thread_context(unsigned n_thread) {
...
+struct null_recipe: arb::recipe {
+    arb::cable_cell_global_properties properties;
+    null_recipe() { properties.default_parameters = arb::neuron_parameter_defaults; }
+    arb::cell_size_type num_cells() const override { return 1; }
+    arb::cell_kind get_cell_kind(arb::cell_gid_type) const override { return arb::cell_kind::cable; }
+    arb::util::unique_any get_cell_description(arb::cell_gid_type) const override { return {arb::cable_cell{}}; }
+    std::vector<arb::probe_info> get_probes(arb::cell_gid_type gid) const override { return {}; }
+    std::any get_global_properties(arb::cell_kind) const override { return properties; }
+};
+
```
Comment (bcumming) [inline test/unit/test_simulation.cpp:50]: A short comment like:
```C++
// Test that cell models with empty morpologies behave as expected,
// i.e. trivial model builds and runs without error.
```

File: test/unit/test_span.cpp [hunk 4]
```diff
@@ -53,11 +53,28 @@ TEST(span, int_iterators) {
...
+        for (auto _: s) ++touched;
+        EXPECT_EQ(s.size(), touched);
+        EXPECT_GT(touched, 0ul); //
+    }
+
+    {
+        auto touched = 0ul;
+        auto s = span(a, a);
+        for (auto _: s) ++touched;
+        EXPECT_EQ(s.size(), touched);
```
Comment (bcumming) [inline test/unit/test_span.cpp:76]: We expect `s.size()==0` here? How about explicitly testing that `touched==0u` to assert that.

**Review by bcumming — CHANGES_REQUESTED** [review]
Good catch.
There are a few suggested comments, and maybe more SFINAE shenanigans required for the span implementation.

File: arbor/backends/multicore/matrix_state.hpp [hunk 5]
```diff
@@ -58,13 +58,15 @@ struct matrix_state {
...
         auto n = size();
         invariant_d = array(n, 0);
-        for (auto i: util::make_span(1u, n)) {
-            auto gij = face_conductance[i];
-
-            u[i] = -gij;
-            invariant_d[i] += gij;
-            if (p[i]!=-1) {
-                invariant_d[p[i]] += gij;
+        if (n >= 1) {
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/backends/multicore/matrix_state.hpp:61]: Sure.

File: arbor/backends/multicore/matrix_state.hpp (hunk 1 above)
Comment (thorstenhater) [reply arbor/backends/multicore/matrix_state.hpp:121]: Sure.

File: arbor/util/span.hpp (hunk 2 above)
Comment (thorstenhater) [reply arbor/util/span.hpp:32]: As `span<K> = range<counter<K>>`, `K` must be both `std::integral` and `std::common_type<I, J>`. Consequently all helpers forward to `make_span(L, R)`.

File: test/unit/test_simulation.cpp (hunk 3 above)
Comment (thorstenhater) [reply test/unit/test_simulation.cpp:50]: Sure.

File: test/unit/test_span.cpp (hunk 4 above)
Comment (thorstenhater) [reply test/unit/test_span.cpp:76]: sure.

File: arbor/backends/multicore/matrix_state.hpp (hunk 1 above)
Comment (thorstenhater) [reply arbor/backends/multicore/matrix_state.hpp:121]: Question: @halfflat mentioned that empty partitions might arise due to other reasons. This check incidentally covers the 'no CV' case, but will also skip those other cases. Therefore, this comment might be misleading.

File: arbor/util/span.hpp (hunk 2 above)
Comment (halfflat) [reply arbor/util/span.hpp:32]: Is it necessary to have this qualification here, when `counter` itself has an `is_integral` guard? Is that guard not working somehow?

Comment (thorstenhater) [reply arbor/util/span.hpp:32]: No, the guard is fine, but I think it's better style to fail here, since then we communicate to the user of `make_span` that `common_type<A, B>` is not `integral`, rather some `C` (which happens to be `common_type<A,B>`). Makes it a bit easier to diagnose those kinds of errors.

**Review by brenthuisman — APPROVED** [review]
I see a small bugfix with tests showing it fixes the bugs + a review that was addressed by the author. I think we're GTG.

File: arbor/util/span.hpp [hunk 6]
```diff
@@ -22,24 +22,24 @@ namespace util {
...
 template <typename I>
 span<I> make_span(I right) {
-    return span<I>(I{}, right);
+    return make_span(I{}, right);
 }
 
 template <typename Seq>
 auto count_along(const Seq& s) {
-    return util::make_span(std::size(s));
+    return make_span(0ul, std::size(s));
```
Comment (halfflat) [inline arbor/util/span.hpp:42]: This potentially changes the return type to something other than `decltype(std::size(s))`, or worse, there may not be a common type between unsigned long and the size type. The original version is correct.

**Review by halfflat — CHANGES_REQUESTED** [review]
Just the one bug.

File: arbor/util/span.hpp (hunk 6 above)
Comment (thorstenhater) [reply arbor/util/span.hpp]: Right.
